### PR TITLE
perf: restore `3.00` wishlist with cached candidates

### LINKED ISSUE #6535: Performance regression from `3.00` to `4.x` when generating pieces to request
### What is the issue?

This issue is about optimising the process of calcuating the list of pieces to be sent out in a BT request message to other peers, and it is next on my to-do list.

In `3.00`, when downloading torrents, Transmission has a cache (namely `tr_swarm::pieces`) storing the list of pieces that

1. are part of a file that is selected for download, and
2. we don't have a complete copy of.

This is a sensible optimisation. Without it, calculating if a piece satisfies these conditions requires sifting through 2 bitfields, and that is expensive. OTOH, these 2 conditions don't change very often, so a caching the results has a large benefit.

However, this optimisation was dropped in 073c6af1d6b5eae29010654150ee269108da719e when converting the code to C++, and this produces flame graphs like this, where `tr_peerMgrGetNextRequests()` is taking up up to half of the CPU time:

flame-4-deb-btf

Compared with a `3.00` flamegraph, the difference is very obvious:

flame-3-deb

I'm already working to add back this optimisation, just writing this issue here for the record. If anyone have any ideas for new optimisations, please also raise them here.

### Which application of Transmission?

None

### Which version of Transmission?

0ce4adf07ce4b6cfabf81e2d027f96197bb9a305

### PR DESCRIPTION
**Pull request by tearfur**
Fixes #6535, more explanation can be found in the GitHub issue.

This PR basically reintroduces the wishlist implementation in `3.00`, which caches piece candidates, and prefers rarer pieces (`4.0.x` wishlist doesn't do any of these).

If your download speed in `4.0.x` has been bottlenecked by CPU, and if my test results don't lie, then you should notice speed improvements in this PR.

Attached a flamegraph captured in a debug build, I captured many other flamegraphs in this PR, and they look about the same as this one:

flame

Notes: Improved libtransmission code to use less CPU.

### CODE REVIEW

**Comment by tearfur** [comment]
Pardon the large diff size, even when not counting the tests. I figured the best thing to do was just to finish the thing in 1 go, since it is basically a rewrite.

**Comment by tearfur** [comment]
**Note to reviewers**

I'm looking for second opinions for how to decouple the wishlist from the peer manager.

So I'm open to doing major refactors to this PR if someone has a better idea.

**Comment by GaryElshaw** [comment]
I don't have better suggestions, but i'm applauding what you're trying to do.

**Comment by killemov** [comment]
@tearfur Great job! I went through the diffs somewhat diagonally and could not find coupling with the peer manager there. I probably overlooked this. But does this mean which peers have what pieces is in the wishlist? I'm not sure about what is wrong with that. Less CPU is nice, but did you look at the MEM impact?

**Comment by tearfur** [comment]
> could not find coupling with the peer manager there

That's what I intended, I'm just wondering if it could be done better.

> but did you look at the MEM impact?

I'm pretty sure it won't be much, even with a torrent with *a lot* of pieces (say 10k), it will be a couple dozen KiB of extra memory memory consumption.

File: libtransmission/peer-mgr-wishlist.cc [hunk 1]
```diff
@@ -14,164 +15,312 @@
...
-        , n_blocks_missing{ missing_in }
-        , priority{ priority_in }
-        , salt{ salt_in }
+    if (std::empty(blocks))
     {
+        return {};
     }
 
-    [[nodiscard]] constexpr auto compare(Candidate const& that) const noexcept // <=>
+    auto spans = std::vector<tr_block_span_t>{};
```
Comment (ckerr) [inline libtransmission/peer-mgr-wishlist.cc:31]: _(nit)_ we can prealloc with `spans.reserve(std::size(blocks));`

**Review by ckerr — APPROVED** [review]

File: libtransmission/peer-mgr-wishlist.cc (hunk 1 above)
Comment (tearfur) [reply libtransmission/peer-mgr-wishlist.cc:31]: 5524f03

**Comment by reardonia** [comment]
Just wanted to thank you @tearfur for this deep analysis and rewrite. the swarm cache was among the thorniest parts of Tr3 and glad to see it back.

**Comment by tearfur** [comment]
@reardonia Thanks. This PR brings one of the biggest performance improvements between `4.0.x` and `4.1.0`, so I'm proud of it.

I've also spotted a couple areas of improvement. Especially `Wishlist::resort_piece()`, it can be *much* faster than it is now. PR should be up by tomorrow.

**Pull request merged**